File: Restante/app/src/main/kotlin/com/marin/catfeina/features/personagens/data/repository/PersonagensRepository.kt
/*
 *  Projeto: Catfeina
 *  Arquivo: PersonagensRepository.kt
 *
 *  Direitos autorais (c) 2025 Marin. Todos os direitos reservados.
 *
 *  Autores: Luiz Carlos Marin / Ivete Gielow Marin / Caroline Gielow Marin
 *
 *  Este arquivo faz parte do projeto Catfeina.
 *  A reprodução ou distribuição não autorizada deste arquivo, ou de qualquer parte
 *  dele, é estritamente proibida.
 *
 *  Nota:
 *
 */


package com.marin.catfeina.features.personagens.data.repository

import app.cash.sqldelight.coroutines.asFlow
import app.cash.sqldelight.coroutines.mapToList
import com.marin.catfeina.core.data.AppDispatchers
import com.marin.catfeina.sqldelight.CatfeinaDatabaseQueries
import com.marin.catfeina.sqldelight.Personagens
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flowOn
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Repositório para gerenciar os dados dos Personagens.
 */
@Singleton
class PersonagensRepository @Inject constructor(
    private val queries: CatfeinaDatabaseQueries,
    private val dispatchers: AppDispatchers
) {

    /**
     * Retorna um Flow com a lista de todos os personagens, ordenados por nome.
     */
    fun getAllPersonagens(): Flow<List<Personagens>> {
        return queries.getAllPersonagens()
            .asFlow()
            .mapToList(dispatchers.io)
            .flowOn(dispatchers.io)
    }
}

File: lib/app/data/repositories/personagens_repository.dart
/*
Projeto: Catfeina
Arquivo: personagens_repository.dart

Direitos autorais (c) 2025 Marin. Todos os direitos reservados.

Autores: Luiz Carlos Marin / Ivete Gielow Marin / Caroline Gielow Marin

Este arquivo faz parte do projeto Catfeina.
A reprodução ou distribuição não autorizada deste arquivo, ou de qualquer parte
dele, é estritamente proibida.

Descrição:
Implementação concreta do repositório de personagens. Utiliza o PersonagemDao
para interagir com o banco de dados.
*/

import 'package:catfeina/app/data/repositories/i_personagens_repository.dart';
import 'package:catfeina/app/data/sources/app_database.dart';
import 'package:catfeina/app/data/sources/daos/personagem_dao.dart';

class PersonagensRepository implements IPersonagensRepository {
  final PersonagemDao _personagemDao;

  PersonagensRepository(this._personagemDao);

  @override
  Future<void> deletePersonagem(int id) {
    return _personagemDao.deletePersonagem(id);
  }

  @override
  Future<Personagem?> getPersonagemPorId(int id) {
    return _personagemDao.getPersonagemPorId(id);
  }

  @override
  Future<List<Personagem>> getTodosPersonagens() {
    return _personagemDao.getTodosPersonagens();
  }

  @override
  Future<void> upsertPersonagem(PersonagensCompanion personagem) {
    return _personagemDao.upsertPersonagem(personagem);
  }

  @override
  Future<List<Personagem>> searchPersonagens(String query) {
    return _personagemDao.searchPersonagens(query);
  }
}
